### ISSUE TITLE
JSON serializer's solution to running out of memory, causing it to request more memory.

### ISSUE THREAD

**Issue by robertlipe** [PlummersSoftwareLLC/NightDriverStrip#413]
## Bug report
JSON serializer essentially monopolizes the system, probably until it crashes in some weird way.
**Problem**

    File file = SPIFFS.open(fileName);

    if (file)
    {
        if (file.size() > 0)
        {
            debugI("Attempting to read JSON file %s", fileName.c_str());

            if (bufferSize == 0)
                bufferSize = std::max((size_t)JSON_BUFFER_BASE_SIZE, file.size());

            // Loop is here to deal with out-of-memory conditions
            while(true)
            {
                pJsonDoc.reset(new AllocatedJsonDocument(bufferSize));

                DeserializationError error = deserializeJson(*pJsonDoc, file);

                if (error == DeserializationError::NoMemory)
                {
                    pJsonDoc.reset(nullptr);
                    file.seek(0);
                    bufferSize += JSON_BUFFER_INCREMENT;
debug stuf; continue.

**Steps**
1.  [ Something Bad Happens ] 
2.  File gets. opened. Size > 0. We enter the while (forever) loop.
3.  It's our first time here. Buffer size is non-zero and presumably BASE_SIZE.
4. .pJsonDoc = malloc(bufferSize) (more or less
5. We try to deserialize it, but for whatever reason (see number 1) error == NoMemory.) I 
6. We reset JsonDoc, presumably to try to free any previous contents. and we rewind.
7. Whatever bufferSize is, now it's more.
8. Go back to top of loop.

So if we failed at 100, we now add BUFFER_INC and try it at 100+BUFFER_INT. That's going to fail, so now we try it at 100 + 2*BUFFER_INC. Eventually, the heat of the processor causes the system to burst into flames.

Now we didn't actually test pJsonDoc for a failure to allocate; we tested dSerialJSON's ability to parse. That's a bit suspect. If the new had actually failed, we'd have gotten an exception, right? (Maybe we're built without exceptions....) 

On a busier system, it's possible that the outage was a temporary thing and trying it again with an attempt to allocate MORE might be a sane thing, but this just runs away.

**Example**
It looks like BUFFER_INCREMENT is 2K.

(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 16459776 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 16461824 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 16463872 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 16465920 bytes

The system is still trying to run. If you read really really fast, you can catch the .1Hz town crier announcement come around:
 of memory serializing object - increasing buffer to 318681088 bytes
W) (SerializeWithBufferSize)(C0) Out of memory serW) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasiPSRAM Free: 0/0, LED FPS: 30 LED B(SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 318685184 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 318687232 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 318689280 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 318691328 bytes
ight: 100%, LED Watts: 11, Buffer: 0/50W) 0, CPU: 004%, 002%, FreeDraw: 0.017

[ ... ] 
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 357695488 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 357697536 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 357699584 bytes
W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 357701632 bytes

Maybe this does something interesting at 2^30 or 2^31, but I'm pretty sure it's not going to actually allocate 357MB on this device with, a couple of physical MB (2-4-8) of ram and no VM. It didn't work last time and it won't work this time.

**Notes**
Somehow, that loop needs an upper bound before declaring failure to the caller or rate limiting or an intervention that if allocating X in this system didn't work, allocating X + 2K is unlikely to work no matter how many times you politely ask.

I agree that the "something bad" should be studied and made to not happen, but we could get a JSON file that's 2GB or something (maybe SPIFFS was corrupted) or the system could just be temporarily low on memory or whatever, but I think the "maybe we opened a large file" case needs a logic rethinking here.

If I just punch the button, it doesn't reek of starting with a GB file or something. It starts quite reasonably.

(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 4096 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 6144 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 8192 bytes
(W) (SerializeWithBufferSize)(C0) Out of memory serializing object - increasing buffer to 10240 bytes

This looks like there's some kind of corruption and deserializeJson is returning a 'no more memory' in case where trying to get more memory isn't going to succeed. The funny thing is that we likely freed and released 4K, 6k, 8k, 10k, and so on until it DID turn into an unreasonably large requested and started failing for a different reason, but kept going in the same loop with the same effect.

**Comment by rbergen** [comment]
This bug report is based on a misunderstanding of the meaning of `DeserializationError::NoMemory`. It does not mean the system ran out of memory, but that the deserialization requires more _deserialization buffer memory_ than was initially allocated in the `new AllocatedJsonDocument()`. How much memory is required for deserialization is near impossible to predict - it depends on the contents of the JSON that's being deserialized, thereby representing a catch 22/chicken-and-egg problem. Therefore, the only solution when the allocated buffer turns out to be too small is increasing the buffer size, and starting over. 

When serializing, the problem is exactly the same, just with the translation happening in the other direction. (And in fact, the uncertainty is even bigger, because you don't even have the JSON file size to go on.)
There, the "out of buffer memory" condition is indicated by certain ArduinoJson calls returning `false` instead of `true`; this is propagated up in the serialization logic implementation within NightDriverStrip. 
**_This is why SerializeToJSON() should_ always _return `true`,_ except _when it receives a `false` from one of the ArduinoJSON serialization functions that signals the out-of-buffer-memory condition by returning `false` itself._**

If you want to learn more about how ArduinoJson goes about (de)serializing JSON, including the fixed allocation of working memory, please refer to the [ArduinoJson documentation](https://arduinojson.org/v6/doc/). I'd recommend starting with the deserialization and serialization tutorials at the top of that page. 

If there is a problem, in my view it is that ArduinoJson doesn't allow its working buffer to grow while (de)serializing. That's not something we can fix.

All we can do is take different starting values and/or increases for the serialization buffers if we want to serialize what we do. That's (only) sensible when we have a stable set of effects and introduces the risk of over-allocating.
The only way to decrease the amount of memory needed is serializing less.

And yes, if at any point a 357701632 buffer is deemed necessary to serialize the effect list, there is a serious issue in the serialization logic implemented in (one/some of) the effect(s) - my guess would be that some SerializeToJSON() function is returning `false` instead of `true`, or that there's an endless loop somewhere. 
Any upper buffer memory limit we set is arbitrary and can bite us in the backside later. SPIFFS won't get corrupted because nothing is written until in-memory serialization has succeeded.

I'm going to close this issue as "won't fix", unless there is a pressing reason not to do so.

**Comment by rbergen** [comment]
> Maybe we try until 10x the file size (which might also be bogus) and then throw a fatal exception.  It's possible, however, that we'd have still been blinking though surely at a reduced rate because our debug prints aren't cheap, so giving up at some arbitrary rate (though at some number, well
below 350 MB, it just gets silly) may not be great either.

Due to a combination of two aspects, I don't think that's an option:

- It'll only work once the file is in place.
- The only scenario where the buffer size can run away by a bug in our codebase is when it concerns serialization. 

The first time after a "clearing flash", the device is booted with an empty SPIFFS partition; the file of which we would like to use the size is actually created _by the first successful serialization_.   That means that during serialization we can't rely on the presence of a file to multiply the size of - it may simply not exist yet. So, we have another JSON serialization catch 22.

So yes, I am going to close this. I'll be working on a "developer's getting started guide" soon, and will then keep in mind the fact that the ArduinoJson (de)serialization approach and buffer-related memory errors is not necessarily very intuitive.
